written/weeks4-5.md

```diff
@@ -1,23 +1,87 @@
+3. How does a pipelined processor compare to a single cycle processor for the following terms? Do they go up, down, or stay the same?
+    * amount of power consumed
+    * amount of time consumed
+    * amount of energy consumed
+4. We don't usually see extremely large pipelines. (Intel processors peaked at 31 stages in the Prescott and Cedar Mill microarchitectures around 2006.) Part of the reason is due to hold time. Why does hold time influence pipeline length? (Hint: what changes when we split a 5-stage pipeline into, say, 100 stages?)
+
+### Hazards
+
+5. Another reason why we don't see extremely large pipelines is due to hazards. Why do hazards influence pipeline length? What happens to the hazard and forwarding units in a longer pipeline?
+6. Would pipeline hazards affect your answer to #1 above? Is it possible to get the same CPI if we were executing a more realistic program?
+7. Name the three types of hazards, and a solution to solve some of the problems that each hazard brings.
+8. True or false: forwarding sends the result of an earlier stage in the pipeline to a later stage. (Example: IF comes earlier in the pipeline than ID.)
+9. True or false: forwarding sends the result of an instruction that comes earlier in program order to a instruction that comes later in program order.
+10. Assume a traditional 5-stage pipeline. Answer the following questions using the code block below.
+    1. Indicate all data dependencies between instructions.
+    2. If we didn't implement forwarding, how many NOPs would we need to insert to ensure correct execution? You may assume that writes to the register file occur on the first half of the cycle, and reads occur on the second half.
+    3. Now assume we implemented forwarding. How many NOPs are left?
+    4. Is is possible to reschedule this code to remove any NOPs? If so, how many NOPs are left?
+
+```
+0:    lw   R5, 0(R2)
+4:    lw   R2, 4(R6)
+8:    add  R3, R2, R1
+12:   sw   R3, 20(R9)
+16:   sub  R8, R7, R8
+20:   and  R8, R8, R5
+```
+
+11. Why do we need to flush instructions? How do we implement this in the hardware?
+12. Assume a traditional 5-stage pipeline. Answer the following questions using the code block below.
+    1. Assume that, when we execute this code, `R7 = 72` and `R8 = 228`. Can we move the `sw` before the `lw`? Why or why not?
+    2. Assume we do not have a forwarding unit. Can we move instruction 8 between instructions 0 and 4 to eliminate some of the NOPs? Why or why not?
+```
+0:    add  R3, R1, R2
+4:    bne  R3, R4, 16
+8:    lw   R5, 124(R7)
+12:   sw   R6, -32(R8)
+16:   add  R4, R1, R3
+20:   ...
+```
+
+13. What two pieces of information does a processor need in order to make a branch prediction?
+14. What are the two types of branch predictors? Which one requires more hardware to implement?
```

Comment:
I don't know the answer to this question.... how would you determine which needed more hardware?